src/helpers/mathjax.ts

```diff
@@ -0,0 +1,206 @@
+  );
+};
+
+const typesetMathMLNodes = (root: Element, windowImpl: Window) => () => {
+  const mathMLNodes = findUnprocessedMath(root);
+
+  if (isEmpty(mathMLNodes)) {
+    return;
+  }
+
+  // style the entire document because mathjax is unable to style individual math elements
+  windowImpl.MathJax.Hub.Queue(
+    () => windowImpl.MathJax.Hub.Typeset(root)
+  );
+};
+
+const markLatexNodesRendered = (latexNodes: Element[]) => () => {
+  // Queue a call to mark the found nodes as rendered so are ignored if typesetting is called repeatedly
+  // uses className += instead of classList because IE
+  const result = [];
+  for (const node of latexNodes) {
+    result.push(node.className += ` ${MATH_RENDERED_CLASS}`);
+  }
+};
+
+// Search document for math and [data-math] elements and then typeset them
+function typesetDocument(root: Element, windowImpl: Window) {
+  const latexNodes = findLatexNodes(root);
+
+  windowImpl.MathJax.Hub.Queue(
+    typesetLatexNodes(latexNodes, windowImpl),
+    typesetMathMLNodes(root, windowImpl)
+  );
+}
+
+const resolveOrWait = (root: Element, resolve: () => void, remainingTries = 5) => {
+  if (
+    remainingTries > 0
+    && (findLatexNodes(root).length || findUnprocessedMath(root).length)
+  ) {
+    setTimeout(() => {
+      resolveOrWait(root, resolve, remainingTries - 1);
+    }, 200);
+  } else {
+    resolve();
+  }
+};
+
+const typesetDocumentPromise = (root: Element, windowImpl: Window): Promise<void> => new Promise((resolve) => {
+  typesetDocument(root, windowImpl);
+  windowImpl.MathJax.Hub.Queue(() => {
+    resolveOrWait(root, resolve);
+  });
+});
+
+// memoize'd getter for typeset document function so that each node's
+// typeset has its own debounce
+const getTypesetDocument = memoize((root, windowImpl) => {
+  // Install a debounce around typesetting function so that it will only run once
+  // every Xms even if called multiple times in that period
+  return debounce(typesetDocumentPromise, 100, {
+    leading: true,
+    trailing: false,
+  }).bind(null, root, windowImpl);
+});
+getTypesetDocument.cache = new WeakMap();
+
+// typesetMath is the main exported function.
+// It's called by components like HTML after they're rendered
+const typesetMath = (root: Element, windowImpl = assertWindow() as Window) => {
+  if (!document.getElementById('MathJax-Script')) {
+    startMathJax();
+  }
+
+  const hasMath = root.querySelector(COMBINED_MATH_SELECTOR);
+  // schedule a Mathjax pass if there is at least one [data-math] or <math> element present
+  if (windowImpl && windowImpl.MathJax && windowImpl.MathJax.Hub && hasMath) {
+    return getTypesetDocument(root, windowImpl)();
+  } else if (hasMath) {
+    // There are math elements present but MathJax is still loading/initializing
+    // Push to a temp queue to be processed in a StartupHook in configuredCallback
+    mathJaxLoadingQueue.push(() => getTypesetDocument(root, windowImpl)());
+  }
+
+  return Promise.resolve();
```

Comment:
rex's implementation actually resolves when the math is done typesetting, which helps with debounces, why does this resolve immediately?

Reply:
for this to be useful you'd have to really handle the queued path on line 155, otherwise its just inconsistent